src/index.js

```diff
@@ -1,390 +1,586 @@
-          attributes.oncreate(element)
-        })
-      }
+  return (nextNode.element = element)
+}
 
-      for (var i = 0; i < node.children.length; i++) {
-        element.appendChild(
-          createElement(
-            (node.children[i] = resolveNode(node.children[i])),
-            isSvg
-          )
-        )
-      }
+var createVNode = function(name, props, children, element, key, type) {
+  return {
+    name: name,
+    props: props,
+    children: children,
+    element: element,
+    key: key,
+    type: type
+  }
+}
 
-      for (var name in attributes) {
-        updateAttribute(element, name, attributes[name], null, isSvg)
-      }
-    }
+var createTextVNode = function(text, element) {
+  return createVNode(text, EMPTY_OBJECT, EMPTY_ARRAY, element, null, TEXT_NODE)
+}
 
-    return element
-  }
+var recycleChild = function(element) {
+  return element.nodeType === 3 // Node.TEXT_NODE
+    ? createTextVNode(element.nodeValue, element)
+    : recycleElement(element)
+}
 
-  function updateElement(element, oldAttributes, attributes, isSvg) {
-    for (var name in clone(oldAttributes, attributes)) {
-      if (
-        attributes[name] !==
-        (name === "value" || name === "checked"
-          ? element[name]
-          : oldAttributes[name])
-      ) {
-        updateAttribute(
-          element,
-          name,
-          attributes[name],
-          oldAttributes[name],
-          isSvg
-        )
-      }
-    }
+var recycleElement = function(element) {
+  return createVNode(
+    element.nodeName.toLowerCase(),
+    EMPTY_OBJECT,
+    map.call(element.childNodes, recycleChild),
+    element,
+    null,
+    RECYCLED_NODE
+  )
+}
 
-    var cb = isRecycling ? attributes.oncreate : attributes.onupdate
-    if (cb) {
-      lifecycle.push(function() {
-        cb(element, oldAttributes)
-      })
-    }
-  }
+var patch = function(container, element, lastNode, nextNode, eventProxy) {
+  var lifecycle = []
 
-  function removeChildren(element, node) {
-    var attributes = node.attributes
-    if (attributes) {
-      for (var i = 0; i < node.children.length; i++) {
-        removeChildren(element.childNodes[i], node.children[i])
-      }
+  patchElement(container, element, lastNode, nextNode, lifecycle, eventProxy)
 
-      if (attributes.ondestroy) {
-        attributes.ondestroy(element)
-      }
-    }
-    return element
-  }
+  while (lifecycle.length > 0) lifecycle.pop()()
 
-  function removeElement(parent, element, node) {
-    function done() {
-      parent.removeChild(removeChildren(element, node))
-    }
+  return nextNode
+}
 
-    var cb = node.attributes && node.attributes.onremove
-    if (cb) {
-      cb(element, done)
-    } else {
-      done()
+export var h = function(name, props) {
+  var node
+  var rest = []
+  var children = []
+  var length = arguments.length
+
+  while (length-- > 2) rest.push(arguments[length])
+
+  if ((props = props == null ? {} : props).children != null) {
```

Comment:
More simplified version:

Reply:
@joseluisq Yes, you are very much right! But there's new directive in town. Now, I'm trying to optimize for speed (while still optimizing for small bundles —this is not gonna end good hehe—) so I'm trying to avoid implicit coercion if possible.

Reply:
Awesome!